irony-cdb.el

```diff
@@ -114,6 +128,57 @@ for files that it cannot handle."
 ;; Functions
 ;;
 
+(defun irony-cdb--choose-closest-path (file paths)
+  "Find the \"best\" path in PATHS matching FILE
+
+If any paths in PATHS is belongs to the same directory
+or a subdirectory of file, we disregard other candidates.
+
+For remaining candidates, \"nearest\" is measured as abs. difference
+in path depth.
+- We prefer deeper paths at level +N to those at level -N.
+- If multiple paths are equally good, we return the last one.
+
+Returns nil if paths isn't a list of at least one element.
+"
+  (when (listp paths)
+    (let ((paths (or
+                  ;; if we find a cdb in cwd or below, don't consider other candidates
```

Comment:
I just reread your comment for the 3rd time to make sure,  and I think I've finally been able to repunctuate it it makes sense to me.

> "or below"? care to explain? 

Sure.  The test is `(lambda (x) (string-prefix-p (file-name-directory file) x))` and any path
which lives in or under that directory will match, by definition. I believe the `children-first` test (based on the test you provided)  shows this works.

> to me it looks like we will favor a cdb in the current directory 
> even if irony-cdb-search-directory-list does not specify it as a first element. 

True, that's what I understood you wanted when you said earlier:
> It's just, if I find a member of irony-cdb-search-directory-list **in the current directory** it 
> should have higher priority than a member in .., no matter how deep the member is.
 
> I think we should not special case the current dir and just take the first best match from 
> irony-cdb-search-directory-list, which defaults to . for the best match.

That's exactly what the *first version* I wrote did and I threw that one away because we wanted:  
`/foo/bar/build/cdb`  (found by "build/", **one** level up)
to win over 
`/cdb` (found by ".", **two** levels up)

...when the search begins at `/foo/bar/x.cpp`. See https://github.com/Sarcasm/irony-mode/pull/237#discussion_r41975148.
 
Here's what it was, despite the problem I just mentioned is this what you want?

```
(defun irony-cdb--locate-dominating-file-with-dirs (filename start-path dirs)
  "Generalizes `locate-dominating-file` to search in given subdirectory list"
  (cl-loop for dir in dirs
           for match = (locate-dominating-file
                        start-path
                        (concat dir "/" filename))
           when match
           return (expand-file-name
                   (concat (mapconcat 'file-name-as-directory (list match dir)  "")
                           filename))
           ))

```